llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -12773,8 +12774,30 @@ SDValue RISCVTargetLowering::lowerVPSplatExperimental(SDValue Op,
     Mask = convertToScalableVector(MaskVT, Mask, DAG, Subtarget);
   }
 
-  SDValue Result =
-      lowerScalarSplat(SDValue(), Val, VL, ContainerVT, DL, DAG, Subtarget);
+  SDValue Result;
+  if (VT.getScalarType() == MVT::i1) {
+    if (isa<ConstantSDNode>(Val)) {
+      SDValue TrueV = DAG.getNode(RISCVISD::VMSET_VL, DL, ContainerVT, VL);
+      SDValue FalseV = DAG.getNode(RISCVISD::VMCLR_VL, DL, ContainerVT, VL);
+      // This select instruction will be optimized
+      Result = DAG.getSelect(DL, VT, Val, TrueV, FalseV);
```

Comment:
Nit, if you've casted it to a ConstantSDNode it's pretty easy to check the value:

```suggestion
    if (auto *C = dyn_cast<ConstantSDNode>(Val)) {
      SDValue Result = DAG.getNode(C->getZExtValue() ? RISCVISD::VMSET_VL : RISCVISD::VMCLR_VL, DL, ContainerVT, VL);
```